flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java

```diff
@@ -1513,13 +1517,20 @@ private CompletableFuture<Void> waitForTerminatingJob(
                                                     throwable));
                                 });
 
-        return jobManagerTerminationFuture.thenAcceptAsync(
+        // keep track of the job as outstanding, if not done
+        if (!jobManagerTerminationFuture.isDone()) {
+            submittedAndWaitingTerminationJobIDs.add(jobId);
+        }
+
+        return FutureUtils.thenAcceptAsyncIfNotDone(
+                jobManagerTerminationFuture,
+                getMainThreadExecutor(),
                 FunctionUtils.uncheckedConsumer(
                         (ignored) -> {
+                            submittedAndWaitingTerminationJobIDs.remove(jobId);
```

Comment:
This isnt run if the submission fails, leaking memory and breaking all subsequent job submissions attempts with a fixed job id.